pyfixest/estimation/demean_.py

```diff
@@ -298,3 +302,84 @@ def demean(
 
     success = not not_converged
     return (res, success)
+
+
+def demean_jax(
+    x: np.ndarray,
+    flist: np.ndarray,
+    weights: np.ndarray,
+    tol: float = 1e-08,
+    maxiter: int = 100_000,
+) -> tuple[np.ndarray, bool]:
+    """JAX implementation of the alternating projections algorithm for demeaning."""
+    # Enable float64 precision
+    config.update("jax_enable_x64", True)
+
+    @partial(jax.jit, static_argnums=(3, 4))
+    def _demean_all(x, flist, weights, n_groups, n_factors):
+        """Demeaning loop using JAX control flow."""
+
+        def _demean_one_factor(x_curr, factor_idx):
+            # Get current factor's group IDs
+            factor_ids = jnp.asarray(flist[:, factor_idx], dtype=jnp.int32)
+
+            # Initialize accumulators for each group
+            group_sums = jnp.zeros((n_groups, x_curr.shape[1]))
+            group_weights = jnp.zeros(n_groups)
+
+            # Accumulate weighted sums and weights
+            def _accumulate(i, val):
+                sums, wts = val
+                idx = factor_ids[i]
+                sums = sums.at[idx].add(x_curr[i] * weights[i])
+                wts = wts.at[idx].add(weights[i])
+                return sums, wts
+
+            # Use fori_loop to match Numba's iteration order
+            group_sums, group_weights = lax.fori_loop(
+                0, x_curr.shape[0], _accumulate, (group_sums, group_weights)
+            )
+
+            # Compute weighted means exactly as in Numba
+            group_means = group_sums / (group_weights[:, None] + 1e-15)
```

Comment:
Here I would drop the `1e-15` , I see no reason to add it? In my experience LLMs are very prone to bake in some guardrails to avoid numerical problems that I never see in computational stats, maybe this is a ML thing?

Reply:
changed!